source/server/watchdog_impl.h

```diff
@@ -23,20 +23,21 @@ class WatchDogImpl : public WatchDog {
       : thread_id_(thread_id), timer_interval_(interval) {}
 
   Thread::ThreadId threadId() const override { return thread_id_; }
-  // Accessors for use by GuardDogImpl to tell if the watchdog was touched recently and reset the
-  // count back to 0 when it has.
-  uint32_t touchCount() const { return touch_count_.load(); }
-  void resetTouchCount() { touch_count_.store(0); }
+  // Used by GuardDogImpl determine if the watchdog was touched recently and reset the touch status.
+  bool getTouchedAndReset() { return touched_.exchange(false, std::memory_order_acq_rel); }
 
   // Server::WatchDog
   void startWatchdog(Event::Dispatcher& dispatcher) override;
-  void touch() override { touch_count_.fetch_add(1); }
+  void touch() override {
+    // Set touched_ if not already set.
+    bool expected = false;
+    touched_.compare_exchange_strong(expected, true, std::memory_order_release,
```

Comment:
I'm not an atomics expert, but I'm pretty sure you can use `store` with `memory_order_relaxed` on this one, and use `memory_order_relaxed` on the exchange above.

The whole thing is eventually consistent so I'm not sure any ordering matters and that would probably be faster?

The only thing I'm not sure of is whether this would effect test determinism, however, if you are using locks to serialize the tests that should force consistency of all previous operations.

Reply:
I'm no expert either.  After thinking about it some more, I think you're totally right - relaxed memory order is sufficient here.  Switched to that.